pkg/helpers/struct_test.go

```diff
@@ -0,0 +1,142 @@
+package helpers
+
+import (
+	"encoding/base64"
+	"testing"
+	"time"
+
+	"github.com/elastic/elastic-agent-libs/mapstr"
+	"github.com/elastic/elastic-agent-shipper-client/pkg/proto/messages"
+	"github.com/stretchr/testify/require"
+)
+
+var result *messages.Value
+
+func BenchmarkCustomUnmarshal(b *testing.B) {
+	testStructType := struct {
+		A int
+		B string
+	}{
+		A: 5,
+		B: "test",
+	}
+	testList := mapstr.M{
+		"field1": mapstr.M{
+			"value":          3,
+			"value-str":      "test",
+			"value-list":     []interface{}{1, 2, 3, 4, 5, 6, 7, 8, 9, 10},
+			"value-list-str": []string{"value1", "value2", "value3", "valu4", "value5", "value6"},
+			"value-struct":   testStructType,
+			"value-bytes":    []byte{0xff, 0xff, 0xff, 0xff, 0xff, 0xff},
+			"value-time":     time.Now(),
+			"value-map": map[string]string{
+				"key1": "value1",
+				"key2": "value2",
+				"key3": "value3",
+				"key4": "value4",
+			},
+		},
+	}
+	var r *messages.Value
+	var err error
+	// benchmark performance for the map, which will usually be the most complex type to marshal
+	// there's a handful of different possilbe ways to handle the NewValue conversion, so it's helpful
+	// to have a benchmark in case we decide to adjust this in the future
+	for i := 0; i < b.N; i++ {
+		r, err = NewValue(testList)
+		if err != nil {
+			b.Logf("error: %s", err)
+			b.FailNow()
+		}
+		result = r
+	}
+}
+
+func TestStructValue(t *testing.T) {
```

Comment:
We aren't hitting a few cases in this test or the benchmarks looking at the code coverage with:

```sh
go test ./pkg/helpers/... -coverprofile=cover.tmp
go tool cover -html cover.tmp
```

We are missing the `map[string]interface{}` and `reflect.Slice` cases which both seem important enough to test. Hitting all the primitive number cases is probably easy to do but also unlikely to matter that much.

I am giving this PR extra scrutiny because any problems here mean the shipper just doesn't work at all, or works unnecessarily slowly :)



<img width="964" alt="Screen Shot 2023-02-09 at 2 26 35 PM" src="https://user-images.githubusercontent.com/3466215/217917195-9cf75df8-ced4-4858-9243-70baedcc4c9d.png">